File: 0078-subsets/0078-subsets.cpp
class Solution {
public:

  void helper(vector<int>& nums ,int i,vector<int> &ds,vector<vector<int>> &ans)
    {
        if(i==(nums.size()))
        {
            ans.push_back(ds);
            return ;
        }
        // pick 
        ds.push_back(nums[i]);
        helper(nums,i +1,ds,ans);
        // not pick
        ds.pop_back();
        helper(nums,i+1,ds,ans);
        
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>>  ans={};
        vector<int> ds={};
         helper(nums,0,ds,ans);
         return ans;
    }
};

File: 0078-subsets/0078-subsets.java
class Solution {
static void helper(int[] nums, List<Integer> ds, List<List<Integer>> ans, int index) {
        if (index == nums.length) {
            ans.add(new ArrayList<>(ds));
            return;
        }
        ds.add(nums[index]);
        helper(nums, ds, ans, index + 1);
        ds.removeLast();
        helper(nums, ds, ans, index + 1);
    }

    public static List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> ds = new ArrayList<>();
        helper(nums, ds, ans, 0);
        return ans;

    }
}